### PR TITLE
Make pinot-distribution to build a pinot-all jar and assemble it

### PR DESCRIPTION
**Pull request by fx19880617**
1. build one fat jar with full dependencies in pinot-distribution, called  `pinot-all-${project.version}-jar-with-dependencies.jar` 
2. in pinot-distribution, replace all the jars under `lib/*.jar` with this `pinot-all-${project.version}-jar-with-dependencies.jar`.
3. clean up `pinot-plugins` dependencies and ensure `pinot-spi`/`pinot-common`/`pinot-core` are in provided scope for all the plugins pom files.

Current distribution directory looks like:
```
pinot-distribution/target/apache-pinot-incubating-0.3.0-SNAPSHOT-bin/apache-pinot-incubating-0.3.0-SNAPSHOT-bin
➜ tree .
.
├── DISCLAIMER
├── LICENSE
├── NOTICE
├── bin
│   ├── pinot-admin.sh
│   ├── pinot-ingestion-job.sh
│   ├── pinot-tools.sh
│   ├── query-comparator.sh
│   ├── quick-start-hybrid.sh
│   ├── quick-start-offline.sh
│   ├── quick-start-realtime.sh
│   ├── scan-query.sh
│   ├── star-query-comparator.sh
│   ├── star-tree-index-viewer.sh
│   ├── start-broker.sh
│   ├── start-controller.sh
│   ├── start-server.sh
│   └── stats-generator.sh
├── conf
│   ├── pinot-admin-log4j2.xml
│   ├── pinot-broker-log4j2.xml
│   ├── pinot-controller-log4j2.xml
│   ├── pinot-ingestion-job-log4j2.xml
│   ├── pinot-server-log4j2.xml
│   ├── pinot-tools-log4j2.xml
│   ├── quickstart-log4j2.xml
│   ├── sample_broker_tenant.json
│   ├── sample_offline_table_config.json
│   ├── sample_perf_benchmark.yaml
│   ├── sample_query_comparison.conf
│   ├── sample_realtime_table_config.json
│   ├── sample_schema.json
│   └── sample_server_tenant.json
├── examples
│   ├── batch
│   │   ├── airlineStats
│   │   │   ├── airlineStats_offline_table_config.json
│   │   │   ├── airlineStats_schema.json
│   │   │   ├── ingestionJobSpec.yaml
│   │   │   └── rawdata
│   │   │       └── 2014
│   │   │           └── 01
│   │   │               ├── 01
│   │   │               │   └── airlineStats_data_2014-01-01.avro
│   │   │               ├── 02
│   │   │               │   └── airlineStats_data_2014-01-02.avro
│   │   │               ├── 03
│   │   │               │   └── airlineStats_data_2014-01-03.avro
│   │   │               ├── 04
│   │   │               │   └── airlineStats_data_2014-01-04.avro
│   │   │               ├── 05
│   │   │               │   └── airlineStats_data_2014-01-05.avro
│   │   │               ├── 06
│   │   │               │   └── airlineStats_data_2014-01-06.avro
│   │   │               ├── 07
│   │   │               │   └── airlineStats_data_2014-01-07.avro
│   │   │               ├── 08
│   │   │               │   └── airlineStats_data_2014-01-08.avro
│   │   │               ├── 09
│   │   │               │   └── airlineStats_data_2014-01-09.avro
│   │   │               ├── 10
│   │   │               │   └── airlineStats_data_2014-01-10.avro
│   │   │               ├── 11
│   │   │               │   └── airlineStats_data_2014-01-11.avro
│   │   │               ├── 12
│   │   │               │   └── airlineStats_data_2014-01-12.avro
│   │   │               ├── 13
│   │   │               │   └── airlineStats_data_2014-01-13.avro
│   │   │               ├── 14
│   │   │               │   └── airlineStats_data_2014-01-14.avro
│   │   │               ├── 15
│   │   │               │   └── airlineStats_data_2014-01-15.avro
│   │   │               ├── 16
│   │   │               │   └── airlineStats_data_2014-01-16.avro
│   │   │               ├── 17
│   │   │               │   └── airlineStats_data_2014-01-17.avro
│   │   │               ├── 18
│   │   │               │   └── airlineStats_data_2014-01-18.avro
│   │   │               ├── 19
│   │   │               │   └── airlineStats_data_2014-01-19.avro
│   │   │               ├── 20
│   │   │               │   └── airlineStats_data_2014-01-20.avro
│   │   │               ├── 21
│   │   │               │   └── airlineStats_data_2014-01-21.avro
│   │   │               ├── 22
│   │   │               │   └── airlineStats_data_2014-01-22.avro
│   │   │               ├── 23
│   │   │               │   └── airlineStats_data_2014-01-23.avro
│   │   │               ├── 24
│   │   │               │   └── airlineStats_data_2014-01-24.avro
│   │   │               ├── 25
│   │   │               │   └── airlineStats_data_2014-01-25.avro
│   │   │               ├── 26
│   │   │               │   └── airlineStats_data_2014-01-26.avro
│   │   │               ├── 27
│   │   │               │   └── airlineStats_data_2014-01-27.avro
│   │   │               ├── 28
│   │   │               │   └── airlineStats_data_2014-01-28.avro
│   │   │               ├── 29
│   │   │               │   └── airlineStats_data_2014-01-29.avro
│   │   │               ├── 30
│   │   │               │   └── airlineStats_data_2014-01-30.avro
│   │   │               └── 31
│   │   │                   └── airlineStats_data_2014-01-31.avro
│   │   └── baseballStats
│   │       ├── baseballStats_offline_table_config.json
│   │       ├── baseballStats_schema.json
│   │       ├── ingestionJobSpec.yaml
│   │       └── rawdata
│   │           └── baseballStats_data.csv
│   └── stream
│       ├── airlineStats
│       │   ├── airlineStats_realtime_table_config.json
│       │   ├── airlineStats_schema.json
│       │   ├── docker
│       │   │   └── airlineStats_realtime_table_config.json
│       │   ├── kafka_0.9
│       │   │   └── airlineStats_realtime_table_config.json
│       │   ├── kafka_2.0
│       │   │   └── airlineStats_realtime_table_config.json
│       │   └── sample_data
│       │       ├── airlineStats_data.avro
│       │       ├── airlineStats_data.json
│       │       └── airlineStats_data.orc
│       └── meetupRsvp
│           ├── kafka_0.9
│           │   └── meetupRsvp_realtime_table_config.json
│           ├── kafka_2.0
│           │   └── meetupRsvp_realtime_table_config.json
│           ├── meetupRsvp_realtime_table_config.json
│           └── meetupRsvp_schema.json
├── lib
│   └── pinot-all-0.3.0-SNAPSHOT-jar-with-dependencies.jar
├── licenses
│   ├── LICENSE-angular.txt
│   ├── LICENSE-antlr.txt
│   ├── LICENSE-aopalliance.txt
│   ├── LICENSE-args4j.txt
│   ├── LICENSE-asm.txt
│   ├── LICENSE-azure-data-lake-store-sdk.txt
│   ├── LICENSE-cddl-1.0.txt
│   ├── LICENSE-cddl-1.1.txt
│   ├── LICENSE-codemirror.txt
│   ├── LICENSE-cpl-1.0.txt
│   ├── LICENSE-epl-1.0.txt
│   ├── LICENSE-epl-2.0.txt
│   ├── LICENSE-esotericsoftware.txt
│   ├── LICENSE-foundation.txt
│   ├── LICENSE-handlebars.txt
│   ├── LICENSE-jcabi.txt
│   ├── LICENSE-jcraft.txt
│   ├── LICENSE-jline.txt
│   ├── LICENSE-jopt-simple.txt
│   ├── LICENSE-jquery-datatables.txt
│   ├── LICENSE-jquery.txt
│   ├── LICENSE-js-beautify.txt
│   ├── LICENSE-mimepull.txt
│   ├── LICENSE-normalize.txt
│   ├── LICENSE-paranamer.txt
│   ├── LICENSE-protobuf.txt
│   ├── LICENSE-scala.txt
│   ├── LICENSE-slf4j.txt
│   ├── LICENSE-underscore.txt
│   └── LICENSE-xmlenc.txt
├── plugins
│   ├── pinot-batch-ingestion
│   │   ├── pinot-batch-ingestion-standalone
│   │   │   └── pinot-batch-ingestion-standalone-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-hadoop
│   │   │   └── pinot-hadoop-0.3.0-SNAPSHOT-shaded.jar
│   │   └── pinot-spark
│   │       └── pinot-spark-0.3.0-SNAPSHOT-shaded.jar
│   ├── pinot-file-system
│   │   ├── pinot-adls
│   │   │   └── pinot-adls-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-gcs
│   │   │   └── pinot-gcs-0.3.0-SNAPSHOT-shaded.jar
│   │   └── pinot-hdfs
│   │       └── pinot-hdfs-0.3.0-SNAPSHOT-shaded.jar
│   ├── pinot-input-format
│   │   ├── pinot-avro
│   │   │   └── pinot-avro-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-csv
│   │   │   └── pinot-csv-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-json
│   │   │   └── pinot-json-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-orc
│   │   │   └── pinot-orc-0.3.0-SNAPSHOT-shaded.jar
│   │   ├── pinot-parquet
│   │   │   └── pinot-parquet-0.3.0-SNAPSHOT-shaded.jar
│   │   └── pinot-thrift
│   │       └── pinot-thrift-0.3.0-SNAPSHOT-shaded.jar
│   └── pinot-stream-ingestion
│       └── pinot-kafka-2.0
│           └── pinot-kafka-2.0-0.3.0-SNAPSHOT-shaded.jar
└── query_console
    ├── css
    │   ├── lib
    │   │   ├── codemirror-elegant.css
    │   │   ├── codemirror-solarized.css
    │   │   ├── codemirror.css
    │   │   ├── foundation.css
    │   │   ├── foundation.min.css
    │   │   └── normalize.css
    │   └── pinot.css
    ├── index.html
    └── js
        ├── init.js
        └── lib
            ├── angular-route.min.js
            ├── angular-route.min.js.map
            ├── angular.min.js
            ├── angular.min.js.map
            ├── beautify.js
            ├── codemirror
            │   ├── codemirror.js
            │   ├── javascript.js
            │   ├── ruby.js
            │   ├── sql.js
            │   └── yaml.js
            ├── foundation
            │   ├── foundation.abide.js
            │   ├── foundation.accordion.js
            │   ├── foundation.alert.js
            │   ├── foundation.clearing.js
            │   ├── foundation.dropdown.js
            │   ├── foundation.equalizer.js
            │   ├── foundation.interchange.js
            │   ├── foundation.joyride.js
            │   ├── foundation.js
            │   ├── foundation.magellan.js
            │   ├── foundation.offcanvas.js
            │   ├── foundation.orbit.js
            │   ├── foundation.reveal.js
            │   ├── foundation.slider.js
            │   ├── foundation.tab.js
            │   ├── foundation.tooltip.js
            │   └── foundation.topbar.js
            ├── handlebars.js
            ├── jquery-2.1.3.min.js
            ├── jquery.dataTables.min.js
            ├── underscore-min.js
            ├── underscore-min.map
            └── underscore.js

77 directories, 167 files

```

### CODE REVIEW

**Comment by siddharthteotia** [comment]
Why move to fat jar?

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-standalone/pom.xml [hunk 1]
```diff
@@ -40,16 +40,19 @@
       <groupId>org.apache.pinot</groupId>
       <artifactId>pinot-spi</artifactId>
       <version>${project.version}</version>
+      <scope>provided</scope>
```
Comment (kishoreg) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-standalone/pom.xml]: we already specify this in root

**Comment by kishoreg** [comment]
> Why move to fat jar?

Goal is to be able to launch any command using the following line

```java -jar -Dplugin.dir=<dir> -Dplugin.includes=<...> pinot-jar-with-dependencies.jar <class_name> <args>```

Another reason is to be able to launch the spark, hadoop jobs using the same syntax

spark-submit  "plugin stuff" pinot-jar-with-dependencies.jar 

without having a fat jar, one has to specify all the libraries and it makes it hard to launch any command especially with spark/hadoop etc.

Also, we were already generating multiple fat jars - pinot-tools-with-dependencies, pinot-common, pinot-spark, pinot-hadoop etc. This is reducing it to just 1 jar.

Let me know if this does not make sense.

**Comment by siddharthteotia** [comment]
> > Why move to fat jar?
> 
> Goal is to be able to launch any command using the following line
> 
> `java -jar -Dplugin.dir=<dir> -Dplugin.includes=<...> pinot-jar-with-dependencies.jar <class_name> <args>`
> 
> Another reason is to be able to launch the spark, hadoop jobs using the same syntax
> 
> spark-submit "plugin stuff" pinot-jar-with-dependencies.jar
> 
> without having a fat jar, one has to specify all the libraries and it makes it hard to launch any command especially with spark/hadoop etc.
> 
> Also, we were already generating multiple fat jars - pinot-tools-with-dependencies, pinot-common, pinot-spark, pinot-hadoop etc. This is reducing it to just 1 jar.
> 
> Let me know if this does not make sense.

I like the idea. Just to be on the same page, the overhead on the user side (whoever is invoking the command) is to write a shell script that first adds the libraries to the classpath and then use the classpath as part of java -cp <classpath> main_class_name args?

**Comment by kishoreg** [comment]
No, we can generate the scripts using the app-assembler plugin. They just have to pass the plugin dir and plugin names as arguments. It’s more about simplifying the way we think about Pinot jar vs plugin jar

**Pull request merged**